### PR TITLE
Fyne send page Implementation and receive page change

### PR DESCRIPTION
**Pull request by metaclips**
This PR implements send page and adds custom button widget and Box layout with varied space sizes to godcr-fyne, solves https://github.com/raedahgroup/godcr/issues/451 https://github.com/raedahgroup/godcr/issues/440 https://github.com/raedahgroup/godcr/issues/422

Screenshot 2019-12-19 at 18 19 38

Screenshot 2019-12-19 at 18 15 16

Screenshot 2019-12-19 at 18 15 28

Screenshot 2019-12-05 at 01 01 51

Former Receive page layout
image
image
New Receive page layout
Screenshot 2019-12-19 at 18 15 41
image
Screenshot 2019-12-19 at 18 23 29

### CODE REVIEW

File: fyne/pages/send.go [hunk 1]
```diff
@@ -1,10 +1,1005 @@
...
+		destinationAddressErrorLabel)
+
+	sendToAccountLabel := canvas.NewText("Send to account", color.RGBA{R: 41, G: 112, B: 255, A: 255})
+	sendToAccountLabel.TextSize = 12
+
+	destinationBox := widget.NewHBox(destinationAddressEntryGroup, selfSendingToAccountGroup, layout.NewSpacer())
+
+	// This hides self sending account dropdown or destination address entry.
+	sendToAccount := widgets.NewClickableBox(widget.NewVBox(sendToAccountLabel), func() {
+		if sendToAccountLabel.Text == "Send to account" {
```
Comment (C-ollins) [inline fyne/pages/send.go]: Use a boolean for this instead

Comment (beansgum) [inline fyne/pages/send.go]: Use a boolean for this instead.
`if selfSendingToAccountGroup.Hidden {`

File: fyne/pages/send.go [hunk 2]
```diff
@@ -1,10 +1,1005 @@
...
+
+	amountErrorLabel = canvas.NewText("", color.RGBA{237, 109, 71, 255})
+	amountErrorLabel.TextSize = 12
+	amountErrorLabel.Hide()
+
+	maxButton := widgets.NewButton(color.RGBA{61, 88, 115, 255}, "MAX", func() {
+		transactionAuthor.UpdateSendDestination(0, temporaryAddress, 0, true)
+		maxAmount, err := transactionAuthor.EstimateMaxSendAmount()
+		if err != nil {
+			amountErrorLabel.Text = "Not enough funds (or not connected)."
```
Comment (beansgum) [inline fyne/pages/send.go]: This error message should be used only for `insufficient_funds` error and "(or not connected)" should be added only if wallet is not synced. Some other errors can also be thrown here.

File: fyne/pages/send.go [hunk 3]
```diff
@@ -1,10 +1,1005 @@
...
+	sendingSelectedAccountLabel        *widget.Label
+	sendingSelectedAccountBalanceLabel *widget.Label
+	sendingSelectedWalletID            int
+
+	Contents *widget.Box
+}
+
+var sendPage sendPageDynamicData
+
+func sendPageContent(multiWallet *dcrlibwallet.MultiWallet, window fyne.Window) (pageContent fyne.CanvasObject) {
```
Comment (beansgum) [inline fyne/pages/send.go]: The entire send page was created in this function, it should be broken down into smaller reusable functions to reduce the possibility of running into a bug. I'd suggest a function each for source account, destination account/address, amount input + fee estimates & lastly total cost + balance after sending.

File: fyne/pages/send.go [hunk 4]
```diff
@@ -1,10 +1,1005 @@
...
+		selectedWallet = multiWallet.WalletWithID(sendPage.sendingSelectedWalletID)
+		accountNumber, err := selectedWallet.AccountNumber(sendPage.sendingSelectedAccountLabel.Text)
+		if err != nil {
+			showErrorLabel("Could not get accounts")
+			log.Println("could not get accounts on account change, reason:", err.Error())
+			return
+		}
+
+		transactionAuthor = selectedWallet.NewUnsignedTx(int32(accountNumber), dcrlibwallet.DefaultRequiredConfirmations)
+		transactionAuthor.AddSendDestination(temporaryAddress, 0, true)
```
Comment (beansgum) [inline fyne/pages/send.go]: This same estimation is repeated on Lines 116 & 117, a function should handle estimation.

File: fyne/pages/send.go [hunk 5]
```diff
@@ -1,10 +1,1005 @@
...
+			refresher(costAndBalanceAfterSendBox, transactionFeeLabel, totalCostLabel, balanceAfterSendLabel, transactionSize)
+			paintedtransactionInfoform.Refresh()
+			sendPage.Contents.Refresh()
+			return
+		}
+
+		transactionAuthor.UpdateSendDestination(0, temporaryAddress, dcrlibwallet.AmountAtom(amountInFloat), false)
+		feeAndSize, err := transactionAuthor.EstimateFeeAndSize()
+		if err != nil {
+			if err.Error() == "insufficient_balance" {
```
Comment (beansgum) [inline fyne/pages/send.go]: Use the dcrlibwallet reference here. `dcrlibwallet.ErrInsufficientBalance`

File: fyne/pages/send.go [hunk 6]
```diff
@@ -1,10 +1,1005 @@
...
+		})
+
+		confirmButton = widgets.NewButton(color.RGBA{41, 112, 255, 255}, "Confirm", func() {
+			confirmButton.Disable()
+			cancelButton.Disable()
+
+			_, err := transactionAuthor.Broadcast([]byte(walletPassword.Text))
+			if err != nil {
+				// do not exit password popup on invalid passphrase
+				if err.Error() == "invalid_passphrase" {
```
Comment (beansgum) [inline fyne/pages/send.go]: `dcrlibwallet.ErrInvalidPassphrase`

**Review by beansgum — CHANGES_REQUESTED** [review]
The following issues need attention.
1. There're a lot of numbers(dimensions) in this file, move `send.go` into its own folder `pages/send/send.go` and also create a new file it `dimensions.go` and declare all your dimensions in it as constants.
2. A global string file should be created to house all the strings used, in this project. Place all strings in the file as constants and import the file wherever strings are needed.
3. Estimates should not be shown if destination address input contains an invalid address.
4. The window size should be the same as the page size.
5. The page title should not be in italics.
6. Put a border around the account selector and remove the line from "From" and "To".
7. Most of the colors used do not match the mockup.

**Review by beansgum — APPROVED** [review]

**Pull request merged**